designs/2022-languages/README.md

```diff
@@ -0,0 +1,644 @@
+
+The `parse()` method receives the file information and information about the ESLint context in order to correctly parse the file. This is a raw parse, meaning that `parse()` should do nothing more than convert the file into an abstract format without doing any postprocessing. This is important so that ESLint can tell how long parsing takes vs. other operations. (One problem with the current `parseForESLint()` approach is that we lose visibility into the true parsing time.)
+
+The `parse()` method is expected to return an object with at least `ast` and `body` properties; the object may also have other properties as necessary for the language to properly create a `SourceCode` object later. The return value from `parse()` is passed directly into `createSourceCode()`.
+
+This method will be called [inside of `Linter`](https://github.com/eslint/eslint/blob/28d190264017dbaa29f2ab218f73b623143cd1af/lib/linter/linter.js#L805) and will require that the language object be passed into the private `parse()` function.
+
+**Async:** For use with current ESLint, `parse()` must be synchronous; ESLint X will allow this method to be asynchronous.
+
+#### The `createSourceCode()` Method
+
+The `createSourceCode()` method is intended to create an instance of `SourceCode` suitable for ESLint and rule developers to use to work with the file. As such, it may perform postprocessing of a parse result in order to get it into a format that is easier to use.
+
+This method will be called [inside of `Linter`](https://github.com/eslint/eslint/blob/28d190264017dbaa29f2ab218f73b623143cd1af/lib/linter/linter.js#L828) and will require that the language object be passed into the private `parse()` function.
+
+**Async:** For use with current ESLint, `createSourceCode()` must be synchronous; ESLint X will allow this method to be asynchronous.
+
+### The `SourceCode` Object
+
+ESLint already has a `SourceCode` class that is used to [postprocess and organize the AST data](https://github.com/eslint/eslint/blob/f89403553b31d24f4fc841424cc7dcb8c3ef689f/lib/source-code/source-code.js#L149) for JavaScript files. This proposal extends the intended purpose of `SourceCode` such that it encapsulates all functionality related to a single file, including traversal. As such, `SourceCode` objects must implement the following interface:
+
+```ts
+interface SourceCode {
+
+    /**
+     * Root of the AST.
+     */
+    ast: ASTNode;
+
+    /**
+     * The body of the file that you'd like rule developers to access.
+     */
+    body: string | ArrayBuffer;
+
+    /**
+     * Traversal of AST.
+     */
+    traverse(): Iterable<TraversalStep>;
+
+    /**
+     * Return all of the inline areas where ESLint should be disabled/enabled.
+     */
+    getDisableDirectives(): Array<DisableDirective>;
+
+    /**
+     * Return any inline configuration provided via comments.
+     */
+    getInlineConfig(): Array<FlatConfig>;
+}
+
+type TraversalStep = VisitTraversalStep | CallTraversalStep;
```

Comment:
If I read the proposal correct, the traversal step stands for either a node visit or a call to specific rules. So a source code object will produce multiple steps for each ast node, right?

One thing I can think of is how to allow a rule to stop being called during traversal. For example, if a rule got matches on a specific function call `foo(foo())`, the rule would like to only report the first foo instead of reporting two foo calls.

I met similar problems when I want to replace code matches that has nested code matches.

Reply:
> If I read the proposal correct, the traversal step stands for either a node visit or a call to specific rules. So a source code object will produce multiple steps for each ast node, right?

Almost. The traversal step is either a node visit or method call. In ESLint currently, we have methods like `onCodePathStart` and `onCodePathEnd` that aren't strictly node traversal but rather intermediate steps in between the traversal. 

The approach in this RFC is designed to allow for that use case.